File: ChicagoTracker/src/fr/cph/chicago/activity/BusActivity.java
/**
 * Copyright 2014 Carl-Philipp Harmant
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package fr.cph.chicago.activity;

import java.io.IOException;
import java.text.ParseException;
import java.util.HashMap;
import java.util.List;
import java.util.Locale;
import java.util.Map;
import java.util.Map.Entry;

import org.apache.commons.collections4.MultiMap;
import org.apache.commons.collections4.map.MultiValueMap;
import org.xmlpull.v1.XmlPullParserException;

import android.app.ActionBar;
import android.app.Activity;
import android.content.ActivityNotFoundException;
import android.content.Intent;
import android.graphics.drawable.Drawable;
import android.net.Uri;
import android.os.AsyncTask;
import android.os.Bundle;
import android.view.Menu;
import android.view.MenuInflater;
import android.view.MenuItem;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.LinearLayout;
import android.widget.TextView;
import android.widget.Toast;
import fr.cph.chicago.ChicagoTracker;
import fr.cph.chicago.R;
import fr.cph.chicago.connection.CtaConnect;
import fr.cph.chicago.connection.CtaRequestType;
import fr.cph.chicago.connection.GStreetViewConnect;
import fr.cph.chicago.data.Preferences;
import fr.cph.chicago.entity.BusArrival;
import fr.cph.chicago.entity.Position;
import fr.cph.chicago.util.Util;
import fr.cph.chicago.xml.Xml;

public class BusActivity extends Activity {

	/** Tag **/
	private static final String TAG = "BusActivity";

	List<BusArrival> busArrivals;

	private String busRouteName;
	private String busRouteId;
	private String bound;
	private Integer busStopId;
	private ImageView streetViewImage, mapImage, directionImage, favoritesImage;
	private Position position;
	private Menu menu;
	private LinearLayout stopsView;
	private boolean firstLoad = true;
	private boolean isFavorite;
	private TextView streetViewText;

	@Override
	protected void onCreate(Bundle savedInstanceState) {
		super.onCreate(savedInstanceState);

		// Load right xml
		setContentView(R.layout.activity_bus);

		busStopId = getIntent().getExtras().getInt("busStopId");
		String busStopName = getIntent().getExtras().getString("busStopName");
		busRouteId = getIntent().getExtras().getString("busRouteId");
		busRouteName = getIntent().getExtras().getString("busRouteName");
		bound = getIntent().getExtras().getString("bound");
		position = new Position();
		position.setLatitude(getIntent().getExtras().getDouble("latitude"));
		position.setLongitude(getIntent().getExtras().getDouble("longitude"));

		this.isFavorite = isFavorite();

		stopsView = (LinearLayout) findViewById(R.id.activity_bus_stops);

		TextView busRouteNameView = (TextView) findViewById(R.id.activity_bus_station_name);
		busRouteNameView.setText(busStopName);

		TextView busRouteNameView2 = (TextView) findViewById(R.id.activity_bus_value);
		busRouteNameView2.setText(busRouteName + " (" + bound + ")");

		streetViewImage = (ImageView) findViewById(R.id.activity_bus_streetview_image);

		streetViewText = (TextView) findViewById(R.id.activity_bus_steetview_text);

		mapImage = (ImageView) findViewById(R.id.activity_bus_map_image);

		directionImage = (ImageView) findViewById(R.id.activity_bus_map_direction);

		favoritesImage = (ImageView) findViewById(R.id.activity_bus_favorite_star);
		if (isFavorite) {
			favoritesImage.setImageDrawable(getResources().getDrawable(R.drawable.ic_save_active));
		}
		favoritesImage.setOnClickListener(new View.OnClickListener() {
			@Override
			public void onClick(View v) {
				BusActivity.this.switchFavorite();
			}
		});

		new DisplayGoogleStreetPicture().execute(position);

		(new LoadData()).execute();

		getActionBar().setDisplayHomeAsUpEnabled(true);
	}

	@Override
	public boolean onCreateOptionsMenu(Menu menu) {
		super.onCreateOptionsMenu(menu);
		this.menu = menu;
		MenuInflater inflater = getMenuInflater();
		inflater.inflate(R.menu.main_no_search, menu);

		ActionBar actionBar = getActionBar();
		actionBar.setNavigationMode(ActionBar.NAVIGATION_MODE_STANDARD);
		actionBar.setDisplayShowTitleEnabled(true);
		actionBar.setTitle("Bus");

		MenuItem refreshMenuItem = menu.findItem(R.id.action_refresh);
		refreshMenuItem.setActionView(R.layout.progressbar);
		refreshMenuItem.expandActionView();

		return true;
	}

	@Override
	public boolean onOptionsItemSelected(MenuItem item) {
		switch (item.getItemId()) {
		case android.R.id.home:
			finish();
			return true;
		case R.id.action_refresh:
			MenuItem menuItem = item;
			menuItem.setActionView(R.layout.progressbar);
			menuItem.expandActionView();
			(new LoadData()).execute();
			Toast.makeText(this, "Refresh...!", Toast.LENGTH_SHORT).show();
			return true;
		}
		return super.onOptionsItemSelected(item);

	}

	private class LoadData extends AsyncTask<Void, Void, List<BusArrival>> {
		@Override
		protected List<BusArrival> doInBackground(Void... params) {
			MultiMap<String, String> reqParams = new MultiValueMap<String, String>();
			reqParams.put("rt", busRouteId);
			reqParams.put("stpid", String.valueOf(busStopId));
			CtaConnect connect = CtaConnect.getInstance();
			try {
				Xml xml = new Xml();
				String xmlResult = connect.connect(CtaRequestType.BUS_ARRIVALS, reqParams);
				return xml.parseBusArrivals(xmlResult);
			} catch (IOException e) {
				e.printStackTrace();
			} catch (XmlPullParserException e) {
				e.printStackTrace();
			} catch (ParseException e) {
				e.printStackTrace();
			}
			return null;
		}

		@Override
		protected void onProgressUpdate(Void... values) {
			// Get menu item and put it to loading mod
			if (menu != null) {
				MenuItem refreshMenuItem = menu.findItem(R.id.action_refresh);
				refreshMenuItem.setActionView(R.layout.progressbar);
				refreshMenuItem.expandActionView();
			}
		}

		@Override
		protected void onPostExecute(List<BusArrival> result) {
			BusActivity.this.busArrivals = result;
			BusActivity.this.buildArrivals();
			if (!firstLoad) {
				MenuItem refreshMenuItem = menu.findItem(R.id.action_refresh);
				refreshMenuItem.collapseActionView();
				refreshMenuItem.setActionView(null);
			}
		}

	}

	private class DisplayGoogleStreetPicture extends AsyncTask<Position, Void, Drawable> {
		private Position position;

		@Override
		protected Drawable doInBackground(Position... params) {
			GStreetViewConnect connect = GStreetViewConnect.getInstance();
			try {
				this.position = params[0];
				return connect.connect(params[0]);
			} catch (IOException e) {
				e.printStackTrace();
				return null;
			}
		}

		@Override
		protected void onPostExecute(Drawable result) {
			int height = (int) getResources().getDimension(R.dimen.activity_station_street_map_height);
			android.widget.RelativeLayout.LayoutParams params = (android.widget.RelativeLayout.LayoutParams) BusActivity.this.streetViewImage
					.getLayoutParams();
			ViewGroup.LayoutParams params2 = BusActivity.this.streetViewImage.getLayoutParams();
			params2.height = height;
			params2.width = params.width;
			BusActivity.this.streetViewText.setText("Street view");
			BusActivity.this.streetViewImage.setLayoutParams(params2);
			BusActivity.this.streetViewImage.setImageDrawable(result);
			BusActivity.this.streetViewImage.setOnClickListener(new View.OnClickListener() {
				@Override
				public void onClick(View v) {
					String uri = String.format(Locale.ENGLISH, "google.streetview:cbll=%f,%f&cbp=1,180,,0,1&mz=1", position.getLatitude(),
							position.getLongitude());
					Intent intent = new Intent(Intent.ACTION_VIEW, Uri.parse(uri));
					intent.setClassName("com.google.android.apps.maps", "com.google.android.maps.MapsActivity");
					try {
						startActivity(intent);
					} catch (ActivityNotFoundException ex) {
						uri = String.format(Locale.ENGLISH, "http://maps.google.com/maps?q=&layer=c&cbll=%f,%f&cbp=11,0,0,0,0",
								position.getLatitude(), position.getLongitude());
						Intent unrestrictedIntent = new Intent(Intent.ACTION_VIEW, Uri.parse(uri));
						startActivity(unrestrictedIntent);
					}
				}
			});
			BusActivity.this.mapImage.setImageDrawable(ChicagoTracker.getAppContext().getResources().getDrawable(R.drawable.da_turn_arrive));
			BusActivity.this.mapImage.setOnClickListener(new View.OnClickListener() {
				@Override
				public void onClick(View v) {
					String uri = "http://maps.google.com/maps?z=12&t=m&q=loc:" + position.getLatitude() + "+" + position.getLongitude();
					Intent i = new Intent(Intent.ACTION_VIEW, Uri.parse(uri));
					i.setClassName("com.google.android.apps.maps", "com.google.android.maps.MapsActivity");
					startActivity(i);
				}
			});
			BusActivity.this.directionImage.setImageDrawable(ChicagoTracker.getAppContext().getResources()
					.getDrawable(R.drawable.ic_directions_walking));
			BusActivity.this.directionImage.setOnClickListener(new View.OnClickListener() {
				@Override
				public void onClick(View v) {
					String uri = "http://maps.google.com/?f=d&daddr=" + position.getLatitude() + "," + position.getLongitude() + "&dirflg=w";
					Intent i = new Intent(Intent.ACTION_VIEW, Uri.parse(uri));
					i.setClassName("com.google.android.apps.maps", "com.google.android.maps.MapsActivity");
					startActivity(i);
				}
			});
			MenuItem refreshMenuItem = menu.findItem(R.id.action_refresh);
			refreshMenuItem.collapseActionView();
			refreshMenuItem.setActionView(null);
			firstLoad = false;
		}
	}

	protected void switchFavorite() {
		if (isFavorite) {
			Util.removeFromBusFavorites(busRouteId, String.valueOf(busStopId), bound, ChicagoTracker.PREFERENCE_FAVORITES_BUS);
			isFavorite = false;
		} else {
			Util.addToBusFavorites(busRouteId, String.valueOf(busStopId), bound, ChicagoTracker.PREFERENCE_FAVORITES_BUS);
			isFavorite = true;
		}
		if (isFavorite) {
			favoritesImage.setImageDrawable(getResources().getDrawable(R.drawable.ic_save_active));
		} else {
			favoritesImage.setImageDrawable(getResources().getDrawable(R.drawable.ic_save_disabled));
		}
	}

	public void buildArrivals() {
		if (busArrivals != null) {
			Map<String, TextView> mapRes = new HashMap<String, TextView>();
			for (BusArrival arrival : this.busArrivals) {
				if (arrival.getRouteDirection().equals(bound)) {
					String destination = arrival.getBusDestination();
					if (mapRes.containsKey(destination)) {
						TextView arrivalView = mapRes.get(destination);
						arrivalView.setText(arrivalView.getText() + " " + arrival.getTimeLeft());
					} else {
						TextView arrivalView = new TextView(ChicagoTracker.getAppContext());
						arrivalView.setText(arrival.getBusDestination() + ": " + arrival.getTimeLeft());
						arrivalView.setTextColor(ChicagoTracker.getAppContext().getResources().getColor(R.color.grey));
						mapRes.put(destination, arrivalView);
					}
				}
			}
			stopsView.removeAllViews();
			for (Entry<String, TextView> entry : mapRes.entrySet()) {
				stopsView.addView(entry.getValue());
			}
		}
	}

	public boolean isFavorite() {
		boolean isFavorite = false;
		List<String> favorites = Preferences.getBusFavorites(ChicagoTracker.PREFERENCE_FAVORITES_BUS);
		for (String fav : favorites) {
			if (fav.equals(busRouteId + "_" + busStopId + "_" + bound)) {
				isFavorite = true;
				break;
			}
		}
		return isFavorite;
	}
}

File: android-app/src/main/kotlin/fr/cph/chicago/core/activity/BusActivity.kt
/**
 * Copyright 2018 Carl-Philipp Harmant
 *
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package fr.cph.chicago.core.activity

import android.annotation.SuppressLint
import android.graphics.Typeface
import android.os.AsyncTask
import android.os.Build
import android.os.Bundle
import android.support.v4.widget.SwipeRefreshLayout
import android.support.v7.widget.Toolbar
import android.widget.ImageView
import android.widget.LinearLayout
import android.widget.TextView
import butterknife.BindColor
import butterknife.BindString
import butterknife.BindView
import butterknife.ButterKnife
import fr.cph.chicago.R
import fr.cph.chicago.core.App
import fr.cph.chicago.core.listener.GoogleMapDirectionOnClickListener
import fr.cph.chicago.core.listener.GoogleMapOnClickListener
import fr.cph.chicago.core.listener.GoogleStreetOnClickListener
import fr.cph.chicago.entity.BusArrival
import fr.cph.chicago.entity.Position
import fr.cph.chicago.entity.enumeration.TrainLine
import fr.cph.chicago.exception.ConnectException
import fr.cph.chicago.exception.ParserException
import fr.cph.chicago.exception.TrackerException
import fr.cph.chicago.service.BusService
import fr.cph.chicago.service.PreferenceService
import fr.cph.chicago.util.Util

/**
 * Activity that represents the bus stop
 *
 * @author Carl-Philipp Harmant
 * @version 1
 */
class BusActivity : AbstractStationActivity() {

    @BindView(R.id.activity_bus_stop_swipe_refresh_layout)
    lateinit var scrollView: SwipeRefreshLayout
    @BindView(R.id.activity_favorite_star)
    lateinit var favoritesImage: ImageView
    @BindView(R.id.activity_bus_stops)
    lateinit var stopsView: LinearLayout
    @BindView(R.id.activity_bus_streetview_image)
    lateinit var streetViewImage: ImageView
    @BindView(R.id.activity_bus_steetview_text)
    lateinit var streetViewText: TextView
    @BindView(R.id.activity_map_image)
    lateinit var mapImage: ImageView
    @BindView(R.id.activity_map_direction)
    lateinit var directionImage: ImageView
    @BindView(R.id.favorites_container)
    lateinit var favoritesImageContainer: LinearLayout
    @BindView(R.id.walk_container)
    lateinit var walkContainer: LinearLayout
    @BindView(R.id.map_container)
    lateinit var mapContainer: LinearLayout
    @BindView(R.id.activity_bus_station_value)
    lateinit var busRouteNameView2: TextView
    @BindView(R.id.toolbar)
    lateinit var toolbar: Toolbar

    @BindString(R.string.bundle_bus_stop_id)
    lateinit var bundleBusStopId: String
    @BindString(R.string.bundle_bus_route_id)
    lateinit var bundleBusRouteId: String
    @BindString(R.string.bundle_bus_bound)
    lateinit var bundleBusBound: String
    @BindString(R.string.bundle_bus_bound_title)
    lateinit var bundleBusBoundTitle: String
    @BindString(R.string.bundle_bus_stop_name)
    lateinit var bundleBusStopName: String
    @BindString(R.string.bundle_bus_route_name)
    lateinit var bundleBusRouteName: String
    @BindString(R.string.bundle_bus_latitude)
    lateinit var bundleBusLatitude: String
    @BindString(R.string.bundle_bus_longitude)
    lateinit var bundleBusLongitude: String
    @BindString(R.string.bus_activity_no_service)
    lateinit var busActivityNoService: String
    @BindString(R.string.request_rt)
    lateinit var requestRt: String
    @BindString(R.string.request_stop_id)
    lateinit var requestStopId: String

    @JvmField
    @BindColor(R.color.grey_5)
    internal var grey_5: Int = 0
    @JvmField
    @BindColor(R.color.grey)
    internal var grey: Int = 0
    @JvmField
    @BindColor(R.color.yellowLineDark)
    internal var yellowLineDark: Int = 0

    private val util = Util
    private val preferenceService = PreferenceService
    private val busService = BusService

    var busArrivals: List<BusArrival> = listOf()
    private lateinit var busRouteId: String
    private lateinit var bound: String
    private lateinit var boundTitle: String
    private var busStopId: Int = 0
    private lateinit var busStopName: String
    private lateinit var busRouteName: String
    private var latitude: Double = 0.0
    private var longitude: Double = 0.0
    private var isFavorite: Boolean = false

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        App.checkBusData(this)
        if (!this.isFinishing) {
            setContentView(R.layout.activity_bus)
            ButterKnife.bind(this)

            busStopId = intent.getIntExtra(bundleBusStopId, 0)
            busRouteId = intent.getStringExtra(bundleBusRouteId)
            bound = intent.getStringExtra(bundleBusBound)
            boundTitle = intent.getStringExtra(bundleBusBoundTitle)
            busStopName = intent.getStringExtra(bundleBusStopName)
            busRouteName = intent.getStringExtra(bundleBusRouteName)
            latitude = intent.getDoubleExtra(bundleBusLatitude, 0.0)
            longitude = intent.getDoubleExtra(bundleBusLongitude, 0.0)

            val position = Position(latitude, longitude)

            isFavorite = isFavorite()

            mapImage.setColorFilter(grey_5)
            directionImage.setColorFilter(grey_5)
            favoritesImageContainer.setOnClickListener { _ -> switchFavorite() }

            favoritesImage.setColorFilter(if (isFavorite) yellowLineDark else grey_5)

            scrollView.setOnRefreshListener { LoadStationDataTask().execute() }
            streetViewImage.setOnClickListener(GoogleStreetOnClickListener(latitude, longitude))
            mapContainer.setOnClickListener(GoogleMapOnClickListener(latitude, longitude))
            walkContainer.setOnClickListener(GoogleMapDirectionOnClickListener(latitude, longitude))

            val busRouteName2 = "$busRouteName ($boundTitle)"
            busRouteNameView2.text = busRouteName2

            // Load google street picture and data
            loadGoogleStreetImage(position, streetViewImage, streetViewText)
            LoadStationDataTask().execute()

            setToolBar()
        }
    }

    private fun setToolBar() {
        toolbar.inflateMenu(R.menu.main)
        toolbar.setOnMenuItemClickListener { _ ->
            scrollView.isRefreshing = true
            LoadStationDataTask().execute()
            false
        }
        util.setWindowsColor(this, toolbar, TrainLine.NA)
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP) {
            toolbar.elevation = 4f
        }
        toolbar.title = busRouteId + " - " + busStopName
        toolbar.setNavigationIcon(R.drawable.ic_arrow_back_white_24dp)
        toolbar.setOnClickListener { _ -> finish() }
    }

    override fun onRestoreInstanceState(savedInstanceState: Bundle) {
        super.onRestoreInstanceState(savedInstanceState)
        busStopId = savedInstanceState.getInt(bundleBusStopId)
        busRouteId = savedInstanceState.getString(bundleBusRouteId)
        bound = savedInstanceState.getString(bundleBusBound)
        boundTitle = savedInstanceState.getString(bundleBusBoundTitle)
        busStopName = savedInstanceState.getString(bundleBusStopName)
        busRouteName = savedInstanceState.getString(bundleBusRouteName)
        latitude = savedInstanceState.getDouble(bundleBusLatitude)
        longitude = savedInstanceState.getDouble(bundleBusLongitude)
    }

    override fun onSaveInstanceState(savedInstanceState: Bundle) {
        savedInstanceState.putInt(bundleBusStopId, busStopId)
        savedInstanceState.putString(bundleBusRouteId, busRouteId)
        savedInstanceState.putString(bundleBusBound, bound)
        savedInstanceState.putString(bundleBusBoundTitle, boundTitle)
        savedInstanceState.putString(bundleBusStopName, busStopName)
        savedInstanceState.putString(bundleBusRouteName, busRouteName)
        savedInstanceState.putDouble(bundleBusLatitude, latitude)
        savedInstanceState.putDouble(bundleBusLongitude, longitude)
        super.onSaveInstanceState(savedInstanceState)
    }

    /**
     * Draw arrivals in current layout
     */
    fun drawArrivals() {
        val tempMap = HashMap<String, MutableList<LinearLayout>>()
        if (busArrivals.isEmpty()) {
            val arrivalView = TextView(applicationContext)
            arrivalView.setTextColor(grey)
            arrivalView.text = busActivityNoService
            val linearLayout = LinearLayout(applicationContext)
            linearLayout.addView(arrivalView)
            tempMap[""] = mutableListOf(linearLayout)
        } else {
            busArrivals
                .forEach { arrival ->
                    val destination = arrival.busDestination
                    if (tempMap.containsKey(destination)) {
                        val oldLayout = tempMap[destination]!!
                        val linearLayout: LinearLayout = oldLayout[oldLayout.size - 1]
                        val arrivalView = TextView(applicationContext)
                        arrivalView.text = if (arrival.isDelay) " Delay" else " " + arrival.timeLeft
                        linearLayout.addView(arrivalView)
                    } else {
                        val linearLayout = LinearLayout(applicationContext)
                        val destinationView = TextView(applicationContext)
                        destinationView.text = destination + ":   "
                        destinationView.setTextColor(grey)
                        destinationView.setTypeface(null, Typeface.BOLD)
                        val arrivalView = TextView(applicationContext)
                        arrivalView.text = if (arrival.isDelay) " Delay" else " " + arrival.timeLeft
                        linearLayout.addView(destinationView)
                        linearLayout.addView(arrivalView)
                        tempMap[destination] = mutableListOf(linearLayout)
                    }
                }
        }
        stopsView.removeAllViews()
        tempMap.entries
            .flatMap { mutableEntry -> mutableEntry.value }
            .forEach { stopsView.addView(it) }
    }

    override fun isFavorite(): Boolean {
        return preferenceService.isStopFavorite(busRouteId, busStopId, boundTitle)
    }

    /**
     * Add or remove from favorites
     */
    private fun switchFavorite() {
        isFavorite = if (isFavorite) {
            preferenceService.removeFromBusFavorites(busRouteId, busStopId.toString(), boundTitle, scrollView)
            favoritesImage.setColorFilter(grey_5)
            false
        } else {
            preferenceService.addToBusFavorites(busRouteId, busStopId.toString(), boundTitle, scrollView)
            preferenceService.addBusRouteNameMapping(busStopId.toString(), busRouteName)
            preferenceService.addBusStopNameMapping(busStopId.toString(), busStopName)
            favoritesImage.setColorFilter(yellowLineDark)
            true
        }
    }

    @SuppressLint("StaticFieldLeak")
    private inner class LoadStationDataTask : AsyncTask<Void, Void, List<BusArrival>>() {

        private var trackerException: TrackerException? = null

        override fun doInBackground(vararg params: Void): List<BusArrival>? {
            try {
                return busService.loadBusArrivals(requestRt, busRouteId, requestStopId, busStopId
                ) { (_, _, _, _, _, _, routeDirection) -> routeDirection == bound || routeDirection == boundTitle }
            } catch (e: ParserException) {
                this.trackerException = e
            } catch (e: ConnectException) {
                this.trackerException = e
            }
            return null
        }

        override fun onProgressUpdate(vararg values: Void) {}

        override fun onPostExecute(result: List<BusArrival>) {
            if (trackerException == null) {
                this@BusActivity.busArrivals = result
                drawArrivals()
            } else {
                util.showNetworkErrorMessage(scrollView)
            }
            scrollView.isRefreshing = false
        }
    }
}
